### PR TITLE
`ADT_C` sorry-free

### PR DESCRIPTION
**Pull request by michaelmcinerney**
This removes all the sorries in the file `ADT_C`, thereby significantly increasing our confidence that the state relation is correct. 

This also removes the `budget` field of scheduling contexts in the executable specification (and Haskell kernel). This brings the executable spec closer to the C code and shifts the condition that the budget is always the sum of all refills into the abstract refinement proofs where the corresponding invariants are more easily available.

### CODE REVIEW

**Comment by Xaphiosis** [comment]
I think it looks mostly fine, I'm confused by "rt crefine: sorries related to cpspace_refill_relation_unique" ... that commit proves some things, changes a def or two, but doesn't introduce any new sorries... what am I missing?

**Comment by michaelmcinerney** [comment]
> I think it looks mostly fine, I'm confused by "rt crefine: sorries related to cpspace_refill_relation_unique" ... that commit proves some things, changes a def or two, but doesn't introduce any new sorries... what am I missing?

That commit clears sorries related to `cpspace_refill_relation_unique`. The sorrying run finished in the previous PR. Should I change the commit to say `clear sorries...` instead?

File: proof/crefine/RISCV64/Detype_C.thy [hunk 1]
```diff
@@ -1693,6 +1702,30 @@ proof -
                      [OF D.valid_untyped tat tlive rl])
     done
 
+  moreover
+  have "\<forall>t \<in> set (ksReleaseQueue s). tcb_at' t s \<and> ko_wp_at' live' t s"
+    apply (insert rlqrun)
```
Comment (lsf37) [inline proof/crefine/RISCV64/Detype_C.thy:1707]: You'd usually write this as
```
from rlqrun
have "..."
```

File: proof/crefine/RISCV64/Detype_C.thy [hunk 2]
```diff
@@ -896,6 +895,16 @@ proof (rule tcb_queue_relation'_cong [OF refl refl refl])
...
 
+lemma tcb_queue_relation_live_restrict:
+  assumes vuc: "s \<turnstile>' capability.UntypedCap d ptr bits idx"
+  and     rel: "\<forall>t \<in> set q. tcb_at' t s"
+  and    live: "\<forall>t \<in> set q. ko_wp_at' live' t s"
+  and      rl: "\<forall>(p :: machine_word) P. ko_wp_at' P p s \<and> (\<forall>ko. P ko \<longrightarrow> live' ko) \<longrightarrow> p \<notin> {ptr..ptr + 2 ^ bits - 1}"
+  shows "tcb_queue_relation' getNext getPrev (cm |` (- Ptr ` {ptr..+2 ^ bits})) q cend chead =
+         tcb_queue_relation' getNext getPrev cm q cend chead"
+  apply (insert assms)
+  by (fastforce simp: tcb_queue_relation'_def tcb_queue_relation_live_restrict')
```
Comment (lsf37) [inline proof/crefine/RISCV64/Detype_C.thy:906]:
```suggestion
  using assms
  by (fastforce simp: tcb_queue_relation'_def tcb_queue_relation_live_restrict')
```

File: proof/refine/RISCV64/SchedContextInv_R.thy [hunk 3]
```diff
@@ -850,7 +850,7 @@ lemma refillNew_corres:
   apply (rule corres_cross_add_guard[where
       Q = "sc_at' sc_ptr and (\<lambda>s'. ((\<lambda>sc. objBits sc = minSchedContextBits + n) |< scs_of' s') sc_ptr)"])
    apply (fastforce dest!: sc_obj_at_cross[OF state_relation_pspace_relation]
-                     simp: obj_at'_def opt_map_red opt_pred_def)
+                     simp: obj_at'_def projectKOs opt_map_red opt_pred_def)
```
Comment (lsf37) [inline proof/refine/RISCV64/SchedContextInv_R.thy:853]: This points to a slight divergence between `master` and `rt`. `projectKOs` should be in the simp set for more recent architectures, in particular for RISCV. Not relevant to this PR, but worth making an issue for, so we don't forget to address it.

Comment (lsf37) [reply proof/refine/RISCV64/SchedContextInv_R.thy:853]: It's strange that the ARM version doesn't need it. It looks like we've added the `projectKO` simps to the set for ARM on `rt` and but not on RISCV64

File: proof/crefine/RISCV64/ADT_C.thy [hunk 4]
```diff
@@ -1065,10 +1080,110 @@ lemma map_to_cnes_eq:
...
+  "\<lbrakk>\<exists>s. ksPSpace s = ah \<and> valid_pspace' s; map_to_scs ah p = Some sc\<rbrakk> \<Longrightarrow>
+   scTCB sc \<noteq> Some 0 \<and> scReply sc \<noteq> Some 0 \<and> scNtfn sc \<noteq> Some 0 \<and> scYieldFrom sc \<noteq> Some 0"
+  apply (clarsimp simp: map_comp_def valid_pspace'_def split: option.splits)
+  by (fastforce simp: valid_obj'_def valid_sched_context'_def)
+
+lemma scRefills_unique:
+  assumes rels: "refill_buffer_relation ah ch gs" "refill_buffer_relation ah' ch gs"
+      and   vs: "ksPSpace s = ah \<and> ksPSpace s p = Some (KOSchedContext sc)"
+      and   vs': "ksPSpace s' = ah' \<and> ksPSpace s' p = Some (KOSchedContext sc')"
+  shows   "scRefills sc = scRefills sc'"
```
Comment (lsf37) [inline proof/crefine/RISCV64/ADT_C.thy:1093]: indent nitpick:

```suggestion
      and   vs: "ksPSpace s = ah \<and> ksPSpace s p = Some (KOSchedContext sc)"
      and  vs': "ksPSpace s' = ah' \<and> ksPSpace s' p = Some (KOSchedContext sc')"
  shows   "scRefills sc = scRefills sc'"
```

File: proof/crefine/RISCV64/ADT_C.thy [hunk 5]
```diff
@@ -1065,10 +1080,110 @@ lemma map_to_cnes_eq:
...
+  apply (rule list_eq_iff_nth_eq[THEN iffD2])
+  apply (drule_tac x=p in spec)+
+  by (fastforce dest: dyn_array_list_rel_pointwise
+                simp: crefill_relation_def refill.expand)
+
+lemma cpspace_sched_context_relation_unique:
+  assumes rels: "cpspace_sched_context_relation ah ch" "cpspace_sched_context_relation ah' ch"
+                "refill_buffer_relation ah ch gs" "refill_buffer_relation ah' ch gs"
+      and   vs: "\<exists>s. ksPSpace s = ah \<and> valid_pspace' s"
+      and   vs': "\<exists>s. ksPSpace s = ah' \<and> valid_pspace' s"
```
Comment (lsf37) [inline proof/crefine/RISCV64/ADT_C.thy:1113]:
```suggestion
  assumes rels: "cpspace_sched_context_relation ah ch" "cpspace_sched_context_relation ah' ch"
                "refill_buffer_relation ah ch gs" "refill_buffer_relation ah' ch gs"
      and   vs: "\<exists>s. ksPSpace s = ah \<and> valid_pspace' s"
      and  vs': "\<exists>s. ksPSpace s = ah' \<and> valid_pspace' s"
```

File: proof/crefine/RISCV64/ADT_C.thy [hunk 6]
```diff
@@ -1065,10 +1080,110 @@ lemma map_to_cnes_eq:
...
+lemma ksPSpace_valid_pspace_replyRefs_nonzero:
+  "\<lbrakk>\<exists>s. ksPSpace s = ah \<and> valid_pspace' s; map_to_replies ah p = Some reply\<rbrakk> \<Longrightarrow>
+   replyTCB reply \<noteq> Some 0 \<and> replyPrev reply \<noteq> Some 0 \<and> replyNext_of reply \<noteq> Some 0"
+  apply (clarsimp simp: map_comp_def valid_pspace'_def split: option.splits)
+  by (fastforce simp: valid_obj'_def valid_reply'_def)
+
+lemma cpspace_reply_relation_unique:
+  assumes rels: "cpspace_reply_relation ah ch" "cpspace_reply_relation ah' ch"
+      and   vs: "\<exists>s. ksPSpace s = ah \<and> valid_pspace' s"
+      and   vs': "\<exists>s. ksPSpace s = ah' \<and> valid_pspace' s"
```
Comment (lsf37) [inline proof/crefine/RISCV64/ADT_C.thy:1156]:
```suggestion
      and   vs: "\<exists>s. ksPSpace s = ah \<and> valid_pspace' s"
      and  vs': "\<exists>s. ksPSpace s = ah' \<and> valid_pspace' s"
```

**Review by lsf37 — APPROVED** [review]
Excellent, this is a big chunk of additional work and brings a lot more confidence to the state relation and proof setup. Also aligns C and spec more closely, which is always good.

Only some minor style nitpicks in the comments that would be good to fix up before merging. For the commit message Raf mentioned, `clear sorries ..` is a good idea and would work.

**Comment by Xaphiosis** [comment]
> > I think it looks mostly fine, I'm confused by "rt crefine: sorries related to cpspace_refill_relation_unique" ... that commit proves some things, changes a def or two, but doesn't introduce any new sorries... what am I missing?
> 
> That commit clears sorries related to `cpspace_refill_relation_unique`. The sorrying run finished in the previous PR. Should I change the commit to say `clear sorries...` instead?

Let me look at the history for a sec so I don't give bad advice... OK, Gerwin does say things like `aarch64 ainvs: asid_map sorries in ArchFinalise` but when he does so, he does not leave it as just one line, following it up with `Close some of the more hairy asid_map-related sorries in ArchFinalise_AI.` There's no standard for what we *do* with sorries, I've seen "prove" (I like that), "clear" (when there are no more left?), "reduce" (when generally reducing sorries in a file).

TLDR: Yes please, something like `rt crefine: clear sorries related to cpspace_refill_relation_unique` (as you suggested) if there are no more of these left, and `rt crefine: prove sorries related to cpspace_refill_relation_unique` if there might be leftovers.

File: proof/refine/RISCV64/SchedContextInv_R.thy (hunk 3 above)
Comment (michaelmcinerney) [reply proof/refine/RISCV64/SchedContextInv_R.thy:853]: `projectKOs` is definitely in the simp set for RISCV64, but not ARM. In this case, it's probably a copying error. I'll remove it.

Comment (michaelmcinerney) [reply proof/refine/RISCV64/SchedContextInv_R.thy:853]: https://github.com/seL4/l4v/issues/687 for the `projectKOs` issue

File: proof/crefine/RISCV64/Refine_C.thy [hunk 7]
```diff
@@ -892,17 +892,20 @@ lemma dmo_domain_user_mem'[wp]:
...
+     (invs' and (\<lambda>s. sch_act_wf (ksSchedulerAction s) s) and ex_abs einvs) \<top>
+     (doUserOp f tc) (doUserOp_C f tc)"
   apply (simp only: doUserOp_C_def doUserOp_def split_def)
   apply (rule corres_guard_imp)
     apply (rule_tac P=\<top> and P'=\<top> and r'="(=)" in corres_split)
        apply (clarsimp simp: simpler_gets_def getCurThread_def
                 corres_underlying_def rf_sr_def cstate_relation_def Let_def)
-      apply (rule_tac P=invs' and P'=\<top> and r'="(=)" in corres_split)
+      apply (rule_tac P="invs' and (\<lambda>s. sch_act_wf (ksSchedulerAction s) s)"
+                  and P'=\<top> and r'="(=)" in corres_split)
```
Comment (corlewis) [inline proof/crefine/RISCV64/Refine_C.thy:904]: I can't remember, did we ever make a decision about how we want to format and indent multi-line rule_tacs? I have a pretty strong preference for the `in ...` part to at least be on its own line, as in my mind it is more important than the variables being instantiated.

**Review by corlewis — APPROVED** [review]

File: proof/crefine/RISCV64/Refine_C.thy (hunk 7 above)
Comment (lsf37) [reply proof/crefine/RISCV64/Refine_C.thy:904]: I'd say this one is admissible, but I'd not object to putting the `in ..` part on a separate line.

File: proof/invariant-abstract/DetSchedAux_AI.thy [hunk 8]
```diff
@@ -630,6 +664,12 @@ lemma valid_sched_tcb_state_preservation_gen:
   assumes sc_refill_cfg2:
     "\<And>P. \<lbrace>\<lambda>s. (\<forall>p. pred_map active_scrc (sc_refill_cfgs_of s) p \<longrightarrow> pred_map P (sc_refill_cfgs_of s) p) \<and> I s\<rbrace>
             f \<lbrace>\<lambda>rv s. \<forall>p. pred_map active_scrc (sc_refill_cfgs_of s) p \<longrightarrow> pred_map P (sc_refill_cfgs_of s) p\<rbrace>"
+  assumes sc_refill_cfg3:
+    "\<And>P. \<lbrace>\<lambda>s. (\<forall>p. pred_map (\<lambda>cfg. \<not> active_scrc cfg) (sc_refill_cfgs_of s) p
+                    \<longrightarrow> pred_map (\<lambda>cfg. scrc_budget cfg = 0) (sc_refill_cfgs_of s) p)
+               \<and> I s\<rbrace>
+          f \<lbrace>\<lambda>_ s. \<forall>p. pred_map (\<lambda>cfg. \<not> active_scrc cfg) (sc_refill_cfgs_of s) p
+                       \<longrightarrow> pred_map (\<lambda>cfg. scrc_budget cfg = 0) (sc_refill_cfgs_of s) p\<rbrace>"
```
[the remark this answers is not in the document]
Comment (lsf37) [reply proof/invariant-abstract/DetSchedAux_AI.thy:672]: Well spotted, we're deferring fixing that one up to a later PR.
